src/protocol_muxer/multiselect/multiselect_instance.cpp

```diff
@@ -0,0 +1,358 @@
+          msg.error(),
+          [wptr = weak_from_this(),
+           round = current_round_](outcome::result<size_t> res) {
+            auto self = wptr.lock();
+            if (self && self->current_round_ == round) {
+              self->onDataWritten(res);
+            }
+          });
+    }
+    send(std::make_shared<MsgBuf>(std::move(msg.value())));
+  }
+
+  void MultiselectInstance::send(Packet packet) {
+    if (is_writing_) {
+      write_queue_.push_back(std::move(packet));
+      return;
+    }
+
+    auto span = gsl::span<const uint8_t>(*packet);
+
+    SL_TRACE(log(), "sending {}", common::dumpBin(span));
+
+    connection_->write(
+        span, span.size(),
+        [wptr = weak_from_this(), round = current_round_,
+         packet = std::move(packet)](outcome::result<size_t> res) {
+          auto self = wptr.lock();
+          if (self && self->current_round_ == round) {
+            self->onDataWritten(res);
+          }
+        });
+
+    is_writing_ = true;
+  }
+
+  void MultiselectInstance::onDataWritten(outcome::result<size_t> res) {
+    is_writing_ = false;
+
+    if (!res) {
+      return close(res.error());
+    }
+    if (!closed_) {
+      if (!write_queue_.empty()) {
+        send(std::move(write_queue_.front()));
+        write_queue_.pop_front();
+        return;
+      }
+
+      if (wait_for_reply_sent_.has_value()) {
+        // reply was sent successfully, closing with success
+        return close(protocols_[wait_for_reply_sent_.value()]);
+      }
+    }
+  }
+
+  void MultiselectInstance::close(outcome::result<std::string> result) {
+    closed_ = true;
+    ++current_round_;
+    write_queue_.clear();
+    Multiselect::ProtocolHandlerFunc callback;
+    callback.swap(callback_);
+
+    owner_.instanceClosed(shared_from_this(), callback, std::move(result));
+  }
+
+  void MultiselectInstance::receive() {
+    if (closed_ || parser_.state() != Parser::kUnderflow) {
+      log()->error("receive(): invalid state");
+      return;
+    }
+
+    size_t bytes_needed = parser_.bytesNeeded();
+
+    assert(bytes_needed > 0);
+
+    if (bytes_needed > kMaxMessageSize) {
+      SL_TRACE(log(), "rejecting incoming traffic, too large message ({})",
+               bytes_needed);
+      return close(ProtocolMuxer::Error::PROTOCOL_VIOLATION);
+    }
+
+    gsl::span<uint8_t> span(*read_buffer_);
+    span = span.first(bytes_needed);
+
+    connection_->read(span, bytes_needed,
+                      [wptr = weak_from_this(), round = current_round_,
+                       packet = read_buffer_](outcome::result<size_t> res) {
+                        auto self = wptr.lock();
+                        if (self && self->current_round_ == round) {
+                          self->onDataRead(res);
+                        }
+                      });
+  }
+
+  void MultiselectInstance::onDataRead(outcome::result<size_t> res) {
+    if (!res) {
+      return close(res.error());
```

Comment:
Minor: more appropriate to transfer error is using `.as_failure()` method
```suggestion
      return close(res.as_failure());
```

Reply:
what is the reasoning?